File: api/viewsets/estudiante.py
#Estudiante View
from django_filters.rest_framework import DjangoFilterBackend
from rest_framework import status, filters, viewsets
from django.db import transaction
from django.contrib.auth.models import User
from rest_framework.decorators import action
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response
from api.permisos import Administrador, CatedraticoUser

from api.models import Estudiante
from api.models import Profile
from api.models import Rol
from api.serializers import EstudianteSerializer
from api.serializers import UserSerializer

class EstudianteViewset(viewsets.ModelViewSet):
    queryset = Estudiante.objects.filter(activo=True).select_related("perfil")
    #definer permiso para este recurso
    permission_classes = [ Administrador | CatedraticoUser ]
 
    filter_backends = (DjangoFilterBackend, filters.SearchFilter, filters.OrderingFilter)
    filter_fields = ("perfil__nombres","perfil__apellidos")
    search_fields = ("perfil__nombres","perfil__apellidos")
    ordering_fields = ("perfil__nombres","perfil__apellidos")

 
    def get_serializer_class(self):
            #Define serializer for API listar
        if self.action == 'list' or self.action == 'retrieve':
            return EstudianteSerializer
    
        
    def create(self, request):
        try:
            with transaction.atomic():
                user = request.user
                datos = request.data

                #validacion de los datos al serializer
                serializer = UserSerializer(data=datos)

                if serializer.is_valid():
                    #insertar los datos luego de validar
                    user = User.objects.create(
                        username = datos.get("username")
                    )
                    #encriptar contraseña
                    usuario = User.objects.get(username=datos.get("username"))
                    usuario.set_password(datos.get("password"))
                    usuario.save()
                    profile = Profile.objects.create(
                        user = user,
                        nombres = datos.get("profile").get("nombres"),
                        apellidos = datos.get("profile").get("apellidos"),
                        direccion = datos.get("profile").get("direccion"),
                        telefono = datos.get("profile").get("telefono"),
                        gender = datos.get("profile").get("gender"),
                        rol = Rol.objects.get(pk = datos.get("profile").get("rol"))
                    )
                    estudiante = Estudiante.objects.create(
                        perfil = profile,
                        carnet = datos.get("profile").get("estudiante").get("carnet"),
                        nombreContacto = datos.get("profile").get("estudiante").get("nombreContacto"),
                        direccionContacto = datos.get("profile").get("estudiante").get("direccionContacto"),
                        telefonoContacto = datos.get("profile").get("estudiante").get("telefonoContacto")
                    )
                else:
                    print("error en la validacion de datos")
            
            return Response({'registro creado'}, status=status.HTTP_200_OK)
        except Exception as e:
            return Response({'detail': str(e)}, status=status.HTTP_400_BAD_REQUEST)
    

    def update(self, request, *args, **kwargs):
        try:
            with transaction.atomic():
                datos = request.data
               
                #Modificar datos Profile
                profile = Profile.objects.get(user=datos.get("id"))
                profile.nombres = datos.get("profile").get("nombres")
                profile.apellidos = datos.get("profile").get("apellidos")
                profile.direccion = datos.get("profile").get("direccion")
                profile.telefono = datos.get("profile").get("telefono")
                profile.gender = datos.get("profile").get("gender")
                profile.save()
                #Modificar datos Estudiante
                estudiante = Estudiante.objects.get(pk=datos.get("idEs"))
                estudiante.nombreContacto = datos.get("profile").get("estudiante").get("nombreContacto")
                estudiante.direccionContacto = direccionContacto = datos.get("profile").get("estudiante").get("direccionContacto")
                estudiante.telefonoContacto = telefonoContacto = datos.get("profile").get("estudiante").get("telefonoContacto")
                estudiante.save()

            return Response({'registro modificado'}, status=status.HTTP_200_OK)
        except Exception as e:
            return Response({'detail': str(e)}, status=status.HTTP_400_BAD_REQUEST)
    
   
    def destroy(self, request, *args, **kwargs):
        try:
            with transaction.atomic():
                #Cambio de estado
                estudiante = self.get_object()
                estudiante.activo = False
                estudiante.save()

                profile = Profile.objects.get(pk = estudiante.perfil_id)
                profile.activo = False
                profile.save()

                user = User.objects.get(pk = profile.user_id)
                user.is_active = False
                user.save()

            return Response({'registro borrado'}, status=status.HTTP_200_OK)
        except Exception as e:
            return Response({'detail': str(e)}, status=status.HTTP_400_BAD_REQUEST)

File: frontend/src/js/redux/modules/estudiante/estudiante.js
import { handleActions } from 'redux-actions';
import { push } from "react-router-redux";
import { initialize as initializeForm } from 'redux-form';
import { NotificationManager } from "react-notifications";
import { api } from "api";

const GUARDAR_LISTADO_ESTUDIANTES = 'GUARDAR_LISTADO_ESTUDIANTES';
const GUARDAR_REGISTRO_ESTUDIANTES = 'GUARDAR_REGISTRO_ESTUDIANTES';

export const listar = () => (dispatch, getStore) => {
    console.log("getStore", getStore());
    api.get('/estudiante').then((response)=>{
        dispatch({ type: GUARDAR_LISTADO_ESTUDIANTES, data: response });
    }).catch((error)=>{
        console.log("error: ", error)
        NotificationManager.error(
            'Ocurrió un error al listar los estudiantes',
            'Error',
            0
        );
    })
}

export const leer = (id) => (dispatch) => {
    api.get(`/estudiante/${id}`).then((response) => {
        dispatch({type: GUARDAR_REGISTRO_ESTUDIANTES, lectura: response });
        dispatch(initializeForm('estudiante', response ));
    }).catch((error)=>{
        console.log("error: ", error)
        NotificationManager.error(
            'Ocurrió un error al consultar el estudiante',
            'Error',
            0
        );
    })
}

export const registroEstudiante = () => (dispatch, getStore) => {
    const data = getStore().form.estudiante.values;
    const formData = {
        username: data.perfil.user.username,
        password: data.password,
        profile: {
            nombres: data.perfil.nombres,
            apellidos: data.perfil.apellidos,
            direccion: data.perfil.direccion,
            telefono: data.perfil.telefono,
            gender: data.perfil.gender,
            rol: data.perfil.rol,
            estudiante: {
                carnet: data.carnet,
                nombreContacto: data.nombreContacto,
                direccionContacto: data.direccionContacto,
                telefonoContacto: data.telefonoContacto
            }
        }
    }
    api.post('/estudiante', formData).then((response) => {
        NotificationManager.success(
            'Estudiante creado',
            'Exito',
            3000
        );
        dispatch(push('/estudiantes'));
    }).catch((error) => {
        console.log("error: ", error)
        NotificationManager.error(
            'Ocurrió un error al registrar el estudiante',
            'Error',
            0
        );
    })
}

export const modificarEstudiante = () => (dispatch, getStore) => {
    const datos = getStore().form.estudiante.values;
    const id = datos.id;
    const formData = {
        id: datos.perfil.user.id,
        idEs: datos.id,
        profile: {
            nombres: datos.perfil.nombres,
            apellidos: datos.perfil.apellidos,
            direccion: datos.perfil.direccion,
            telefono: datos.perfil.telefono,
            gender: datos.perfil.gender,
            rol: datos.perfil.rol,
            estudiante: {
                carnet: datos.carnet,
                nombreContacto: datos.nombreContacto,
                direccionContacto: datos.direccionContacto,
                telefonoContacto: datos.telefonoContacto
            }
        }
    }
    
    api.put(`/estudiante/${id}`, formData).then((response) => {
        NotificationManager.success(
            'Estudiante modificado correctamente',
            'Exito',
            3000
        );
        dispatch(push('/estudiantes'));
    }).catch((error) => {
        console.log("error: ", error)
        NotificationManager.error(
            'Ocurrió un error al modificar el estudiante',
            'Error',
            0
        );
    })
}

export const eliminar = (id) => (dispatch) => {
    api.eliminar(`/estudiante/${id}`).then((response) => {
        NotificationManager.success(
            'Estudiante borrado correctamente',
            'Exito',
            3000
        );
        dispatch(listar());
    }).catch((error) => {
        console.log("error: ", error)
        NotificationManager.error(
            'Ocurrió un error al borrar el estudiante',
            'Error',
            0
        );
    })
}

export const actions = {
    registroEstudiante,
    modificarEstudiante,
    listar,
    leer,
    eliminar,
};

export const reducers = {
    [GUARDAR_LISTADO_ESTUDIANTES]: (state, { data }) => {
        return {
            ...state,
            data,
        };
    },
    [GUARDAR_REGISTRO_ESTUDIANTES]: (state, { lectura }) => {
        return {
            ...state,
            lectura,
        };
    },
};

export const initialState = {
    loader: false,
    data: null,
    lectura: null,
};

export default handleActions(reducers, initialState)